wgsl/index.bs

```diff
@@ -9597,29 +9899,50 @@ value|non-uniform=] during the execution of the function call.
+    * If the function has a [=return type=], create a node called <dfn noexport>Value_return</dfn>.
 * Desugar pointers as described in [[#pointer-desugar]].
-    * For each pointer parameter in the [=address spaces/function=] address space,
+    * For each formal parameter that is a pointer in the [=address spaces/function=] address space,
         create a <dfn noexport>Value_return_i</dfn> node.
+        This represents the function's effect on the uniformity of the value pointed to by the *i*<sup>th</sup> parameter.
 * Walk over the syntax of the function, adding nodes and edges to the graph following the rules of the next sections ([[#func-var-value-analysis]], [[#uniformity-statements]], [[#uniformity-function-calls]], [[#uniformity-expressions]]), using [=CF_start=] as the starting control-flow for the function's body.
-* For each [=Value_return_i=] node, record which [=param_i=] nodes are reachable from it.
-* Look at which nodes are reachable from [=RequiredToBeUniform=].
-    * If this set includes the node [=MayBeNonUniform=], then reject the program.
-    * If this set includes [=CF_start=], then the [=call site tag=] for the function is [=CallSiteRequiredToBeUniform=].
-    * Otherwise, the [=call site tag=] is [=CallSiteNoRestriction=].
-    * For each [=param_i=] in this set, the corresponding [=parameter tag=] is [=ParameterRequiredToBeUniform=].
-    * Remove from the graph all nodes that have been visited.
-* If [=Value_return=] exists, look at which nodes are reachable from it
-    * If this set includes [=MayBeNonUniform=], then the [=function tag=] is [=ReturnValueMayBeNonUniform=].
-    * For each [=param_i=] in this set, the corresponding [=parameter tag=] is [=ParameterRequiredToBeUniformForReturnValue=].
-* For each [=Value_return_i=] node, look at which nodes are reachable from it
-    * If this set includes [=MayBeNonUniform=], the corresponding [=pointer parameter tag=] is [=PointerParameterMayBeNonUniform=].
-    * Otherwise, the corresponding [=pointer parameter tag=] is [=PointerParameterNoRestriction=].
-* If the [=function tag=] has not been assigned, then it is [=NoRestriction=].
-* For each parameter, if it has not been assigned a [=parameter tag=], then it is [=ParameterNoRestriction=].
-
-Note: The entire graph can be destroyed at this point. The tags listed above are all that we need to remember to analyze callers of this function.
+* Initialize as follows:
+    * The [=function tag=] is initialized to [=NoRestriction=].
+    * The [=call site tag=] is initialized to [=CallSiteNoRestriction=].
+    * The [=parameter tag=] for each [=param_i=] is initialized to [=ParameterNoRestriction=].
+    * The [=pointer parameter tag=] for each [=param_i=], if it exists, is initialized to [=PointerParameterNoRestriction=].
+* For each severity *S* in the order {[=severity/error=], [=severity/warning=], [=severity/info=]}, perform the following:
```

Comment:
Previously the analysis removed nodes after traversing from RequiredToBeUniform. That has a knock-on effect with how the rules for function calls are constructed later. Consider the following:

```
fn foo(c : bool) -> bool {
  @diagnostic(info, derivative_uniformity)
  if (c) {
    let x = textureSample(...);
  }
  return c;
}

var<private> x : bool;
fn bar() {
  let tmp = foo(x);
  if (tmp) {
    let s = textureSample(...);
  }
}
```

`foo` gets tagged with `RequiredToBeUniform.info` and parameter `c` gets tagged with `ParameterRequiredToBeUniform.info`. Therefore, in `bar`, the graph is disconnected between before and after the function call to `foo`. 

So there will be an info diagnostic triggered for the sample in `foo`, but it will miss the error diagnostic for the sample in `bar. The easiest change might be to always connect `CF_after` to the last `CF_i`.

Reply:
Yeah, there were two subtle things:
- statefulenss: removing nodes as you go along
- and the early exit implied by the "reject the shader", which now can't just be "reject the shader".

Reply:
The early exit change isn't enough here. Right now (with this PR), the `ParameterRequiredToBeUniform.S` and `ParameterRequiredToBeUniformForReturnValue` states are mutually exclusive, which is why this doesn't work.

In Tint, we split the tag to allow those two states to co-exist ([see here](https://source.chromium.org/chromium/chromium/src/+/627b57a43f94511a0cfdea3aab071282cc2bb64b:third_party/dawn/src/tint/resolver/uniformity.cc;l=146-149)). This also ties in to our solution for #3677, which has a similar problem.